File: DP/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res, min_prod, max_prod = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            min_prod, max_prod = min(min_prod * nums[i], max_prod * nums[i], nums[i]), max(min_prod * nums[i], max_prod * nums[i], nums[i])
            res = max(res, max_prod)
        return res
                    

File: DP/152. Maximum Product Subarray.cpp
// Pay attention to the Intermediate varialbe.
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
        int max_prod = nums[0], min_prod = nums[0], res = nums[0];
        for(int i = 1; i < nums.size(); i++)
        {
            int max_temp = max_prod, min_temp = min_prod;
            max_prod = max(nums[i], max(max_temp * nums[i], min_temp * nums[i]));
            min_prod = min(nums[i], min(max_temp * nums[i], min_temp * nums[i]));
            res = max(max_prod, res);
        }
        return res;
    }
};
